File: easy-solution/minesp3164/206.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        answer = None
        while head:
            temp = head
            head = head.next
            temp.next = answer
            answer = temp
        return answer
"""
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head: return head   #Empty.
        if not head.next: return head   #We reached end.
        orig_head = self.reverseList(head.next)  #Traverse to end, orig_head is now end node.
        head.next.next = head   #Swap head with right node.
        head.next = None   #So we don't wind up in infinite loop.
        return orig_head   #Very last thing returned. End node!
        """

File: easy-solution/galaxy4276/206.java
import java.io.*;

class Solution {
  public ListNode reverseListByRecursive(ListNode head) {
      if (head == null || head.next == null) return head;
      ListNode resultNode = reverseListByRecursive(head.next);
      head.next.next = head;
      head.next = null;
      System.out.println("result Node: " + resultNode.val);
      return resultNode;
  }

  public ListNode reverseListByIteration(ListNode head) {
    ListNode parent = null;
    while (head != null) {
      ListNode curr = head;
      head = head.next;
      curr.next = parent;
      parent = curr;
    }
    return parent;
  }

}

public class Main {
  public static void main(String[] args) {
    Solution s = new Solution();
    ListNode resultNode = s.reverseListByIteration(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5))))));
    while (resultNode != null) {
      System.out.print(resultNode.val + "->");
      resultNode = resultNode.next;
    }
  }

}
